clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1034,28 +1024,35 @@ private Record nextFetchedRecord() {
                     records = currentBatch.streamingIterator();
                 }
 
-                Record record = records.next();
-                maybeEnsureValid(record);
-
-                // skip any records out of range
-                if (record.offset() >= nextFetchOffset) {
-                    nextFetchOffset = record.offset() + 1;
-
-                    // control records are not returned to the user
-                    if (!currentBatch.isControlBatch())
-                         return record;
+                // If an exception was thrown from the last record. We should throw the same exception here again.
+                // We do not cache the exception thrown last time because the stack trace could be different
+                // from last time.
+                if (hasExceptionInLastFetch) {
+                    maybeEnsureValid(lastRecord);
```

Comment:
Yeah, I was wondering about that, too. We do the refetch for replica fetchers on the brokers. I was not doing refetch in the patch mostly due to the additional complexity (backoff time or maximum retries etc). Given that is an improvement for a very edge case and we need this fix for 0.11.0, maybe we can do that in a follow up patch?